### PR TITLE
feature: add publish action

### PR DESCRIPTION
**Pull request by yue4u**
https://github.com/pixiv/three-vrm/issues/1294

## Description

- add publish action via workflow_dispatch
- support npm provenance

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
+name: Publish to npmjs
+
+on:
+  # Who has permission to workflow_dispatch
+  # https://github.com/orgs/community/discussions/26622
+  workflow_dispatch:
```
Comment (yue4u) [inline .github/workflows/publish.yml:6]: ` if: contains('["..."]', github.actor)` does not really make sense to me

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  fetch:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v3
+      with:
+        # checkout is pinned to dev branch here
+        ref: dev
+    - name: Setup Node
+      uses: actions/setup-node@v3
```
Comment (yue4u) [inline .github/workflows/publish.yml:28]: v4 use node 20 we should update CI before that

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+          ${{ runner.os }}-yarn-
+    - name: Fetch Deps
+      run: yarn install --frozen-lockfile
+
+  publish:
+    runs-on: ubuntu-latest
+    needs: fetch
+    permissions:
+      contents: write
+      id-token: write
```
Comment (yue4u) [inline .github/workflows/publish.yml:48]: https://docs.npmjs.com/generating-provenance-statements#publishing-packages-with-provenance-via-github-actions

File: .github/workflows/publish.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    steps:
+    - name: Set git username and email
+      run: |
+        git config --local user.email "action@github.com"
+        git config --local user.name "actions-user"
+
+    - name: Set package version
+      run: |
+        yarn lerna version ${{ github.event.inputs.version }} --exact --no-push --yes
```
Comment (yue4u) [inline .github/workflows/publish.yml:57]: I'm not sure if we need `--force-publish` here

File: .github/workflows/publish.yml [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        yarn lerna publish from-git --dist-tag latest
+      env:
+        NPM_CONFIG_PROVENANCE: true
+        NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
+
+    - name: Git push version up commits
+      run: |
+        git push origin ${{ github.ref_name }}
+
+    - name: Upload Builds
```
Comment (yue4u) [inline .github/workflows/publish.yml:67]: Not sure if we want to automate the release tag on Github. There are only 3rd party solutions for now. https://github.com/actions/create-release#github-action---releases-api

**Review by yue4u — COMMENTED** [review]
Anyway this might not work as intended unless we merge and test it on dev branch.
Current plan is to merge this and iterate through a few canary releases.

File: .github/workflows/publish.yml (hunk 1 above)
Comment (0b5vr) [reply .github/workflows/publish.yml:6]: `workflow_dispatch` is a slightly recent feature iirc; maybe they couldn't use the feature at that moment.
I agree with using `workflow_dispatch` :+1:

File: .github/workflows/publish.yml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    steps:
+    - name: Set git username and email
+      run: |
+        git config --local user.email "action@github.com"
+        git config --local user.name "actions-user"
+
+    - name: Set package version
+      run: |
+        yarn lerna version ${{ github.event.inputs.version }} --exact --no-push --yes
+        yarn lerna publish from-git --dist-tag latest
```
Comment (0b5vr) [inline .github/workflows/publish.yml]: it would be cool if we could make the disttag variable

Comment (yue4u) [reply .github/workflows/publish.yml]: add in 42d5348c99f1c14872d500643622e91cad40b359

**Comment by 0b5vr** [comment]
let's try whether it's working properly

**Review by 0b5vr — APPROVED** [review]

**Pull request merged**